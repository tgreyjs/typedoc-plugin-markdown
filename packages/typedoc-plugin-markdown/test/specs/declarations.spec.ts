import * as Handlebars from 'handlebars';

import { TestApp } from '../test-app';

describe(`Declarations:`, () => {
  let testApp: TestApp;

  let template: Handlebars.TemplateDelegate;

  beforeAll(async () => {
    testApp = new TestApp(['declarations.ts']);
    await testApp.bootstrap();
    TestApp.stubPartials(['member.sources']);
    template = TestApp.getPartial('member.declaration');
  });

  test(`should compile a variable with default value`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('stringWithDefaultValueDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile an undefined declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('undefinedNumberDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('objectLiteralDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal cast as a const`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('objectLiteralAsConstDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type literal declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('typeLiteralDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile declaration with double underscores in name and value`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('__DOUBLE_UNDERSCORES_DECLARATION__'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile any function type`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('AnyFunctionType'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('functionDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile callable declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('callableDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile indexable declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('indexableDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile enum delcaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('EnumDeclarations').children[0],
      ),
    ).toMatchSnapshot();
  });
});
