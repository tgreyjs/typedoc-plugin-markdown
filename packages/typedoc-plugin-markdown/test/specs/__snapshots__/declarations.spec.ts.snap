// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Declarations: should compile a variable with default value 1`] = `
"• \`Const\` **stringWithDefaultValueDeclaration**: \`\`\\"variable\\"\`\` = 'variable'

[partial: member.sources]
"
`;

exports[`Declarations: should compile an undefined declaration 1`] = `
"• \`Let\` **undefinedNumberDeclaration**: \`number\`

[partial: member.sources]
"
`;

exports[`Declarations: should compile any function type 1`] = `
"Ƭ **AnyFunctionType**<A\\\\>: (...\`input\`: \`any\`[]) => \`A\`

#### Type parameters

| Name | Type |
| :------ | :------ |
| \`A\` | \`A\` = \`any\` |

#### Type declaration

▸ (...\`input\`): \`A\`

##### Parameters

| Name | Type |
| :------ | :------ |
| \`...input\` | \`any\`[] |

##### Returns

\`A\`

[partial: member.sources]
"
`;

exports[`Declarations: should compile callable declaration 1`] = `
"• \`Let\` **callableDeclaration**: (\`someArg\`: \`number\`) => \`boolean\`

#### Call signature

▸ (\`someArg\`): \`boolean\`

##### Parameters

| Name | Type |
| :------ | :------ |
| \`someArg\` | \`number\` |

##### Returns

\`boolean\`

#### Type declaration

| Name | Type |
| :------ | :------ |
| \`arg1\` | \`string\` |
| \`arg2\` | \`number\` |

[partial: member.sources]
"
`;

exports[`Declarations: should compile declaration with double underscores in name and value 1`] = `
"• \`Const\` **\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_**: typeof [\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_](../modules.md#__double_underscores_declaration__)

[partial: member.sources]
"
`;

exports[`Declarations: should compile enum delcaration 1`] = `
"• **Down** = \\"DOWN\\"

[partial: member.sources]
"
`;

exports[`Declarations: should compile function declaration 1`] = `
"• \`Let\` **functionDeclaration**: (\`someArg\`: \`number\`) => \`boolean\`

[partial: member.sources]
"
`;

exports[`Declarations: should compile indexable declaration 1`] = `
"• \`Let\` **indexableDeclaration**: \`Object\`

#### Index signature

▪ [index: \`number\`]: \`string\`

#### Type declaration

| Name | Type |
| :------ | :------ |
| \`arg1\` | \`string\` |

[partial: member.sources]
"
`;

exports[`Declarations: should compile object literal cast as a const 1`] = `
"• \`Const\` **objectLiteralAsConstDeclaration**: \`Object\`

Comments

#### Type declaration

| Name | Type | Description |
| :------ | :------ | :------ |
| \`Prop1\` | \`\`\\"Prop1\\"\`\` | Comment for Prop1. |
| \`Prop2\` | \`\`\\"Prop2\\"\`\` | Comment for Prop2. |
| \`Prop3\` | \`\`\\"Prop3\\"\`\` | - |

[partial: member.sources]
"
`;

exports[`Declarations: should compile object literal declaration 1`] = `
"• \`Const\` **objectLiteralDeclaration**: \`Object\`

#### Type declaration

| Name | Type | Description |
| :------ | :------ | :------ |
| \`valueA\` | \`number\` | - |
| \`valueB\` | \`boolean\` | - |
| \`valueC\` | \`Object\` | - |
| \`valueX\` | \`Object\` | description for valueX |
| \`valueX.valueA\` | \`number\`[] | - |
| \`valueX.valueZ\` | \`string\` | - |
| \`valueY\` | () => \`string\` | description for valueY |
| \`valueZ\` | \`string\` | description for valueZ |

[partial: member.sources]
"
`;

exports[`Declarations: should compile type literal declaration 1`] = `
"• \`Let\` **typeLiteralDeclaration**: \`Object\`

#### Type declaration

| Name | Type | Description |
| :------ | :------ | :------ |
| \`valueA?\` | \`number\` | Comments for valueA |
| \`valueB?\` | \`boolean\` | - |
| \`valueX\` | \`Object\` | Comment for valueX |
| \`valueX.valueA\` | \`number\`[] | - |
| \`valueX.valueY\` | (\`z\`: \`string\`) => { \`a\`: \`string\` ; \`b\`: \`string\`  } | - |
| \`valueX.valueZ\` | \`string\` | Nested comment for valueZ |
| \`valueY\` | () => \`string\` | - |
| \`valueZ\` | \`string\` | Comment for valueZ |

[partial: member.sources]
"
`;
